### ISSUE TITLE
[#3164] QC-Py-33-RL-PPO: per-asset reward incoherence — policy figée near-uniform, collapse Buy (val_sharpe 0.657 = buy-and-hold dégénéré)

### ISSUE THREAD

**Issue by jsboige** [jsboige/CoursIA#3359]
## Contexte

Audit de fidelite **#3164** (lot QC-Py-33/34/35). Finding **REINVENTION involontaire** sur `MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-33-RL-PPO-Trading.ipynb`. **Classe distincte** du frozen-agent (#3326 PPO research, #3341 DQN deadlock) : ici la boucle d'entrainement est **atteignable** et l'optimizer recoit des gradients, mais le signal d'avantage est **architecturalement incoherent** -> la policy reste figee near-uniform -> collapse deterministic vers Buy partout -> deguise en buy-and-hold.

## Bug (per-asset reward incoherence, confirme par lecture directe + outputs)

Cell `fd6281ba` remplit le rollout buffer en decomposant un signal **portfolio-level** (un seul scalaire `reward`) sur des critic-values **per-asset** :

```python
for a_idx in range(N_ASSETS):
    asset_state = state[a_idx * 6:(a_idx + 1) * 6]
    rollout_buffer["rewards"].append(reward / N_ASSETS)   # MEME scalaire pour les 15 assets
    rollout_buffer["values"].append(values_list[a_idx])   # mais V(asset_a_state) per-asset
```

Le critic per-asset `V(state_a)` ne peut pas fitter un signal de recompense **identique** pour les 15 assets au meme timestep -> `value_loss` stagne ~0.43-0.57 -> avantages normalises = bruit -> `policy_loss ~= 0` -> parametres a peine perturbes -> entropy figee near-max -> argmax collapse vers Buy sur ce seed/data.

## Preuves outputs committes (G.1 cross-check, non-propage un verdict non verifie)

Cell `fd6281ba` stdout (3 evals a 30720/61440/92160 steps) :
```
Step  30,720 | pol_loss=-0.0026 | val_loss=0.4290 | entropy=1.319 | val_sharpe=+0.657 | best=+0.657
Step  61,440 | pol_loss=-0.0009 | val_loss=0.5724 | entropy=1.299 | val_sharpe=+0.657 | best=+0.657
Step  92,160 | pol_loss=+0.0002 | val_loss=0.5407 | entropy=1.278 | val_sharpe=+0.657 | best=+0.657
```
- `val_sharpe=+0.657` **identique aux 3 evals** (deterministe, pas une policy apprise).
- `entropy 1.319 -> 1.278` sur 60K steps ; max pour 4 actions = ln(4) = **1.386** -> policy near-uniform, a peine bougee.
- `policy_loss ~= 0` a chaque checkpoint -> aucun signal de gradient.

Cell `4587b5d3` (action distribution finale, apres load best model) :
```
Actions - Hold: 0, Buy: 11280, Sell: 0, Close: 0
```
11280 = 15 assets x 752 val steps = **100% Buy** a chaque timestep = buy-and-hold deguise (sur bull market 2022-2024 large-cap US, Sharpe 0.657 / +31.3%).

## Non-fidelite au papier cite (Schulman 2017 PPO)

PPO implemente correctement en isolation (clipping, GAE, value clipping, optimizer.step sur les memes params que forward). Mais la **decomposition per-asset d'une reward portfolio-level** n'est pas dans le papier : PPO operere sur un unique signal (state, action, reward) au niveau ou la reward est definie. Ici la reward est portfolio-level (le env `TradingEnvironment.step` retourne un scalar unique pour le portefeuille), mais on force 15 critic-heads per-asset a l'apprendre -> incoherence. Le verdict "PPO apprend" est faux : la policy est l'init figee.

## Cause racine secondaire (leakage)

Les tickers incluent `AAPL, MSFT, NVDA, AMZN, GOOGL, META` (FAANG/Mag7) sur 2014-2024 -> viole regle C `pr-review-discipline.md` (pas de FAANG/Mag7 en training set). Pas la cause de la degenerescence (le bug per-asset suffit) mais issue de fidelite separée.

## Fix chirurgical propose (anti-regression : restaurer l'intention, pas sorry)

**Option 1 (preferee, architecturalement correcte)** : policy portfolio-level. L'env opere deja au niveau portefeuille (state dim 90 = 15x6, reward scalar). Entrainer UN actor-critic sur l'observation complete, output UNE action portfolio-level par timestep. Cela matche le PPO cite et elimine l'incoherence.
- Edit cell `ActorCritic.__init__` : `state_dim=90` (full state) au lieu de slices per-asset.
- Edit cell `fd6281ba` : passer `state` complet (pas `asset_state` slices) ; ne pas diviser reward par N_ASSETS (un seul (s,a,r) par timestep).

**Option 2 (minimal, moins fidèle)** : si on garde la decomposition per-asset, il faut que le critic per-asset soit entraîné sur un signal per-asset coherent — mais la reward est genuiment portfolio-level, donc c'est architecturalement faux. Option 1 est le vrai fix.

**Bonus** : cell comparison table hardcode `dqn_metrics = {"Sharpe": 0.657}` alors que le vrai DQN (`QC-Py-32`) est Sharpe 0.818, Return -72.5%, action dist Hold:2760/Buy:545/Sell:768/Close:7 (politique diverse reelle). Corriger le baseline ou retirer la ligne.

## Transparence G.1

- **Confiance HAUTE** sur le verdict REINVENTION (triple signal entropy-flat + all-Buy + val_sharpe identique = non-ambigu).
- **Confiance HAUTE** que la boucle d'entrainement est atteignable (pas le pattern #3341) — verifie par l'arithmetique rollout/step.
- **Confiance MOYENNE** sur le mecanisme exact (per-asset GAE incoherence) : explication la plus consistante avec l'empreinte entropy-stuck + value_loss=0.5, mais non re-confirme par re-execution avec le fix. **Le reviewer/coordinator doit re-verifier en re-executant avec Option 1 et en checkant que entropy < ~1.0 et val_sharpe varie entre evals.**

`See #3164`

**Comment by jsboige** [comment]
## [DIAGNOSTIC NUANCE — à lire AVANT merge de la PR #3376] Le Sharpe 0.657 est l'optimum buy-and-hold, pas (seulement) un artefact du bug critic

En investiguant #3360 (A2C, même classe de bug), j'ai découvert que le **diagnostic original de cette issue était partiellement incomplet**. Partage pour transparence G.1/G.2 avant que ai-01 ne merge #3376.

### Preuve empirique (policies fixes sur val 2022-2024)
```
all-Hold   : return=+0.0%   sharpe=+0.000
all-Buy    : return=+31.3%  sharpe=+0.657   <-- valeur exacte produite par PPO buggy ET PPO fixed
all-Sell   : return=-59.1%  sharpe=-1.982
all-Close  : return=+0.0%   sharpe=+0.000
```
Sur train 2014-2021 : `all-Buy = +392% / sharpe +1.156` (bull market massif FAANG/Mag7).

### Ce que ça signifie
1. Le bug per-asset (critic incohérent) était **réel** — la décomposition reward/N_ASSETS sur 15 critic-values per-asset est architecturalement fausse.
2. **MAIS** le Sharpe 0.657 n'était **pas causé** par le bug critic. C'est la **stratégie optimale triviale** pour une policy portfolio-level single-action sur un bull market large-cap US. `all-Buy` produit exactement 0.657/+31.3%.
3. **Toute** policy portfolio-level single-action correctement entraînée converge vers all-Buy — c'est l'optimum réel (train reward +1.156 pousse massivement vers Buy).
4. Le fix portfolio-level (#3376, Option 1) est **architecturalement correct** (critic cohérent, PPO canonique), MAIS **ne changera PAS le verdict Sharpe** — la policy apprend all-Buy *honnêtement* au lieu de l'*accident* via le bug.

### Re-évaluation honnête de la PR #3376 (que j'ai livrée cycle 8)
Mon [DONE] cycle 8 disait « val_sharpe varie 0.645/0.162/0.089 → non-dégénéré ». **C'était trop optimiste.** La variation pendant le training était de l'**overfitting au-delà de l'optimum all-Buy** (la policy s'éloigne du buy-and-hold puis se dégrade), pas de la diversification réelle. Le verdict final (Sharpe 0.645) **EST** all-Buy. Le bug REINVENTION est résolu (la policy apprend au lieu d'être figée), mais le **résultat** reste buy-and-hold car c'est l'optimum.

### Décision pour ai-01
- **Merger #3376 ?** Oui si on valorise la fidélité architecturale (la policy devrait apprendre all-Buy *honnêtement* via un critic cohérent, pas *par accident* via un critic cassé). Le fix est correct, il rend juste un verdict identique pour une meilleure raison.
- **Revert ?** Non — le fix est valide, c'est ma *caractérisation du résultat* qui était imparfaite, pas le code.
- **Lever le vrai problème :** le verdict « NO BEATS = buy-and-hold » est **inhérent** à la conception portfolio-level single-action + bull market FAANG. Pour qu'un RL trading notebook montre une vraie valeur ajoutée, il faudrait soit (a) per-asset actions différenciées, (b) exclure les FAANG/Mag7 (règle C), (c) une période avec drawdowns réels. C'est un scope plus large qu'un fix de bug.

`See #3164` `See #3360`

**Comment by jsboige** [comment]
Resolved by PR #3376 (merged, squash ea645b73d). The per-asset critic-incoherence REINVENTION is fixed — single portfolio-level actor-critic, the policy now trains honestly (entropy moves, policy_loss non-zero, val_sharpe varies) instead of being frozen. Verdict is NO BEATS for the correct reason now (all-Buy is the trivial large-cap-bull optimum, not a frozen disguise). The deeper redesign track (per-asset differentiated actions / FAANG-Mag7 exclusion / drawdown period) stays tracked in #3360. See #3164.
